File: Python/15_3Sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()

        sol_set = set()
        for i in range(len(nums) - 1):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            l = i + 1
            r = len(nums) - 1
            while l < r:
                summed = nums[i] + nums[l] + nums[r]

                if summed == 0:
                    tup = tuple([nums[i], nums[l], nums[r]])
                    if tup not in sol_set:
                        sol_set.add(tup)
                    l += 1
                elif summed > 0:
                    r -= 1
                else:
                    l += 1


        return list(sol_set)

File: C++/15_3_Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        std::set<std::tuple<int, int, int>> set;
        for(int i = 0; i < nums.size() - 2;i++) {
            int l = i + 1;
            int r = nums.size() - 1;
            while(l < r) {
                int sum = nums[i] + nums[l] + nums[r];

                if(sum == 0) {
                    set.insert({nums[i], nums[l], nums[r]});
                    
                    l++;
                } else if(sum < 0) {
                    l++;
                } else {
                    r--;
                }
            }
        }

        std::vector<std::vector<int>> sums;
        for (auto [a, b, c] : set) {
            sums.push_back({a,b,c});
        }
        return sums;
    }
};
